Iceberg-Parquet Storage Strategies for OCSF events (#138)
ocsf/ocsf-docs: articles/OCSF-Storage-Strategies.md

Reviewer: It's worth noting that within Iceberg's hidden partitioning scheme, evolving a partition only affects data going forward.

Author: I think I covered that in the Schema Evolution section:

"The changes take place and are realized for data written after the changes or until rewriting or compaction takes place"

# OCSF – Iceberg Storage Strategies
Paul Agbabian, April 2026; Updated May, June 2026

> Contributions by Matthias Valentin, Hunter Madison

Reviewer 2: ```suggestion
Contributions by Matthias Vallentin, Hunter Madison
```
(Double "l". Thanks for the credit!)

## Overview
There are multiple ways that OCSF structured events can be stored for analysis in Parquet/Iceberg table format. The following list is not exhaustive but forms the most common approaches seen so far in the industry. 

1.  Single table
2.	By event source
3.	By event category
4.	By event class
5.	By Observable and category

> 6.  By Required only

Reviewer 2: The list mixes tabs and spaces, which renders unevenly, and item 6 has no corresponding section below (see review body for a proposed draft).
```suggestion
1. Single table
2. By event source
3. By event category
4. By event class
5. By Observable and category
6. By Required only
```

### Parquet
Parquet is a column-oriented file format whose files hold column values contiguously, unlike row-oriented file formats like CSV or JSON, where each file stores columns across rows, where rows are stored contiguously. Due to the diversity of column types, and distribution of column values, this traditional approach does not compress as efficiently, as values need to be skipped across the columns of each row. For low cardinality columns, the compressibility is highest. Just as important, search performance tends to favor contiguously stored column values since most projections and predicates do not require all the columns in a row. Therefore I/O load is greatly reduced, and the time required to return a resultset is optimized.

Iceberg doesn't define its own file format, but rather supports different underlying physical file formats, including Parquet. Avro, another format supported by Iceberg, is an example of a row-oriented format.

### Partitioning

> Partitioning of a table is a physical organization of the underlying files, based on combining similar rows based on column values. This helps speed queries by only having to read the files that are needed for a given query. For example, partitioning based on time, say by day, means queries that are outside of a time range only need to open files in the paritions that span the day range of the query. Multiple partitions are possible, for example, within a `day` partition, the rows (that is, the column files of rows) can be organized by `product` keeping the events from specific products together within each day.

Reviewer 2: ```suggestion
Partitioning of a table is a physical organization of the underlying files, based on combining similar rows based on column values. This helps speed queries by only having to read the files that are needed for a given query. For example, partitioning based on time, say by day, means queries that are outside of a time range only need to open files in the partitions that span the day range of the query. Multiple partitions are possible, for example, within a `day` partition, the rows (that is, the column files of rows) can be organized by `product` keeping the events from specific products together within each day.
```
("paritions" → "partitions")

> Iceberg has hidden partitioning and schema evolution features that make table maintenance easier than with Hive or traditional RDBMS tables. Partitions can be made by values that are not explicit columns. For example, if a column is `time` with a millisecond resolution, Iceberg can parition by day without having to include a `day` column. And partitions can change over time, as with schema. Nevertheless, maintenance and onboarding of new sources are dynamic considerations for any of the approaches discussed.

Reviewer 2: ```suggestion
Iceberg has hidden partitioning and schema evolution features that make table maintenance easier than with Hive or traditional RDBMS tables. Partitions can be made by values that are not explicit columns. For example, if a column is `time` with a millisecond resolution, Iceberg can partition by day without having to include a `day` column. And partitions can change over time, as with schema. Nevertheless, maintenance and onboarding of new sources are dynamic considerations for any of the approaches discussed.
```
("parition" → "partition")

#### Example Table

| Attribute | Parquet Column Type | Comment |
| --------- | ----------- | ------- |
├ time            |   bigint  |   ← Base required, partition on *days(time)*
├ class_uid       |   int     |   ← Base required, partition on *class_uid*
├ category_uid    |   int     |   ← Base required, partition on *category_uid*
├ activity_id     |   int     |   ← Base required
├ severity_id     |   int     |   ← Base required
├ metadata        |   struct<version, product struct<name,vendor_name>, ...>    |   ← Base required, partition on *product.name*
├ observables     |   list<struct<name, type, value, ...>>  |   ← Base optional
├ device          |   struct<hostname, os struct<name,type>, ...>  |  ← [some class attribute], null for events that don't have one
├ actor           |   struct<user struct<name,uid>, process struct<...>, ...>   | ...
├ src_endpoint    |   struct<ip, port, ...> | ...
├ dst_endpoint    |   struct<ip, port, ...> | ...
├ file            |   struct<name, hashes list<struct<...>>, ...>   | ...

> └── ... (the rest of the join across classes)

Reviewer 2: The tree-drawing characters (`├`, `└──`) inside table rows break GitHub's table rendering: rows lose their leading pipe and the last line isn't a table row at all. Two clean options: a proper Markdown table (below), or a fenced code block to keep the tree aesthetic. Also, the types shown are Iceberg/SQL type names rather than Parquet physical types, so I'd relabel the column.
```suggestion
| Attribute | Iceberg Type | Comment |
| --------- | ------------ | ------- |
| `time` | `bigint` | Base required, partition on `days(time)` |
| `class_uid` | `int` | Base required, partition on `class_uid` |
| `category_uid` | `int` | Base required, partition on `category_uid` |
| `activity_id` | `int` | Base required |
| `severity_id` | `int` | Base required |
| `metadata` | `struct<version, product struct<name, vendor_name>, ...>` | Base required, partition on `product.name` |
| `observables` | `list<struct<name, type, value, ...>>` | Base optional |
| `device` | `struct<hostname, os struct<name, type>, ...>` | Class-specific; null for events that don't have one |
| `actor` | `struct<user struct<name, uid>, process struct<...>, ...>` | ... |
| `src_endpoint` | `struct<ip, port, ...>` | ... |
| `dst_endpoint` | `struct<ip, port, ...>` | ... |
| `file` | `struct<name, hashes list<struct<...>>, ...>` | ... |
| ... | ... | the rest of the union across classes |
```

### Schema Evolution

> Iceberg allows for schema column changes, partitioning and sorting changes at any time after tables are defined and populated. The changes take place and are realized for data written after the changes or until rewriting or compaction takes place. This mitigates some of the table mainteance that is necessary when either new products are onboarded or the OCSF schema itself evolves from version to version.

Reviewer 2: ```suggestion
Iceberg allows for schema column changes, partitioning and sorting changes at any time after tables are defined and populated. The changes take place and are realized for data written after the changes or until rewriting or compaction takes place. This mitigates some of the table maintenance that is necessary when either new products are onboarded or the OCSF schema itself evolves from version to version.
```
("mainteance" → "maintenance")

### Compaction

> Over time, data can get stale, or the frequency of new data subsides such that the underlying files become too small, hindering compression, and forcing more file opens. For partitioning and sorting changes that would benefit historical data, rewriting the table files to compact them into more optimum sizes (usually around 125 MB) can be done. This is termed compaction and is part of the maintenance of the Iceberg lakehouse. As schema changes can be made at any time, for example adding class tables, new partitions, or adding columns to existing tables, the volume flow frequency that fragments the Parquet files will benefit from compaction periodically.

Reviewer 2: Small correction on the number: the conventional Parquet file/row-group target is 128 MB, and Iceberg's own default write target (`write.target-file-size-bytes`) is 512 MB. Suggest either citing 128 MB or a range.
```suggestion
Over time, data can get stale, or the frequency of new data subsides such that the underlying files become too small, hindering compression, and forcing more file opens. For partitioning and sorting changes that would benefit historical data, rewriting the table files to compact them into more optimum sizes (typically 128 MB to 512 MB; Iceberg's default write target is 512 MB) can be done. This is termed compaction and is part of the maintenance of the Iceberg lakehouse. As schema changes can be made at any time, for example adding class tables, new partitions, or adding columns to existing tables, the volume flow frequency that fragments the Parquet files will benefit from compaction periodically.
```

Compaction does not need to be applied to an entire table, but can be applied to specific partitions or date ranges.

### Limits

> There are no theoretical limits in Iceberg as to how many columns or rows a given table can contain, and with schema evolution, alterations are metadata changes that don't need to touch existing data files. However in practice, too many columns in particular can surface operational issues depending on query engines, catalogs and field ID assignments. The latter is of particular concern with OCSF and Iceberg, due to the number of unique columns the array and structure decomposition entails.

Reviewer: It's worth mentioning that while Iceberg doesn't have any defined limits, the query engine will.

Author: I think that was covered in the Limits section:

"Query engines like AWS Athena can fail with more than 7500 columns, ..."

### OCSF Profiles

The structure of Iceberg tables is directly a function of the OCSF attributes of the class or classes modeled by the table. OCSF profiles can augment those attributes to broaden the applicability of a class, and those profile attributes are not bound by the categories of the classes. For example, adding `device` and `actor` attributes to network classes can be done by applying the `Host` profile to `network_activity`. This would be common for EDR events that can report network activity from a computer, including what user or process initiated the activity.

This means that the Iceberg tables will have additional columns from the profiles that may be applied to classes. These columns can be added a priori, ready for events that include profile attributes, or as part of the schema evolution of the tables, when onboarding event sources from products that emit them.

## Single Table Join-Union

> One obvious approach that has been used at large scale is a single, partitioned table with structured Parquet columns. Logically, separate class tables would be joined (as with a LEFT OUTER JOIN), and the table would be a union of the events across all classes.

Reviewer 2: "LEFT OUTER JOIN" implies a join key and row matching, which isn't what happens here, since each event populates exactly one class's columns. I think what's meant is a union of the class *schemas* (shared Base columns merged once, class-specific columns null for other classes).
```suggestion
One obvious approach that has been used at large scale is a single, partitioned table with structured Parquet columns. Logically, the table's schema is the union of all class schemas: the shared Base class attributes appear once, while class-specific attributes are null for events of other classes. The table then holds the union of all events across all classes.
```

Given the number of OCSF dictionary attributes that can be combined into the number of OCSF objects, and optional profiles that add attributes across classes and objects, the distinct number of underlying Iceberg column Field IDs can be very high. Nevertheless, with proper partitioning this approach has its advantages. In practice, not every class is required for the particular event sources that are stored. They can be added as necessary with schema evolution metadata updates.

### Single Table Pros
-	A stable number of tables (1)
-	Reuse of table by multiple products
-	Minimize duplication of related columns within a table
-   No cross-table joins for all use cases including single source or product queries
-   Single target for all ETL targets
-   Leverage schema evolution for new class and partition maintenance

### Single Table Cons
-   Extremely high number of columns may hit metadata and query engine operational limits

> -   The union of all events creates a massive dataset, impacting compaction

Reviewer: I would say that a single table limits what you can do in terms of partitioning, not compaction. The union of two parquet files is a cheap operation.

Author: Got it. I'll make that change.

-	Table holds all related classes in a domain
-	Reduce the number of joins within a domain’s category
-	ETL is straightforward as category_uid directs the events to its destination

### Category Cons
-	Searching for all events from a single product requires searching all tables if the product events span categories
-   Table will be larger than by source or by event class impacting compaction

## By Event Class
Each OCSF event is an instance of a single OCSF class. Events across different sources share the same classes, distinguished by their `metadata.source` or `metadata.product.name` attributes.

However, there are at this time of writing more than 50 event classes and new classes are developed more often than new categories of classes. Hence to cover a wide range of products that may share some but not all of the same classes, many tables need to be maintained and periodically added.

More joins are needed when analysis spans classes, as they map to separate tables. Given the commonality of many of the attributes across classes, columnar compression is not maximized. However, for targeted analysis, tables will be smaller (fewer rows).

### Class Pros

> •	ETL is straightforward as `class_uid` directs the events to its destination

Reviewer 2: ```suggestion
-	ETL is straightforward as `class_uid` directs the events to its destination
```
(`•` → `-` to match the rest of the list.)

This strategy is hybrid. It takes advantage of an OCSF schema feature called Observables.

Observables are an array of the most common attributes surfaced from any of the events that are emitted in a flat generic array. Each element of the array is a single Observable whose `value` attribute holds the element data. The `name` attribute provides the schema context of the attribute, for example `src_endpoint.ip` and `dst_endpoint.ip` are distinct attributes of the `ip` observable, and the `class_uid` and `activity_id` are present for behavior.

The original intent of the observables was for threat intelligence matching, where the total fidelity and context of the event was not required, just the observable values that might match Indicators of Compromise (IoCs).

There is another use case for Observables: a table constructed for observables will be common across every event class, and therefore every event across all products. This table can have foreign keys to dimension tables much like a STAR schema in OLAP. The Observables array is flattened into about 40 columns, along with the most important Base event attribute columns that identify and classify the events.

> In practice, many analytics can run very efficiently directly against a single table across all products and classes, while drill-down and detailed investigation is performed by a minimum number of joins to dimension tables (i.e. joins are not required for detection analytics, only for drill-down).

Reviewer 2: Carrying over my note from #137: this holds for value-match detections (IOC/retro matching, the original purpose), but detections that depend on attribute *context or relationships* beyond what `observables[].name` plus the Base columns capture still need the dimension tables. Suggest scoping the parenthetical.
```suggestion
In practice, many analytics can run very efficiently directly against a single table across all products and classes, while drill-down and detailed investigation is performed by a minimum number of joins to dimension tables. Value-match detections, such as IOC matching and retrospective hunting over observable values, need no joins at all; detections that depend on richer event context than the observables and Base attributes carry will still join to the dimension tables.
```

Author: > Carrying over my note from #137: this holds for value-match detections (IOC/retro matching, the original purpose), but detections that depend on attribute _context or relationships_ beyond what `observables[].name` plus the Base columns capture still need the dimension tables. Suggest scoping the parenthetical.

I thought I had updated this. Note that the context of the observable is always available via the `name` attribute: 
`The full name of the observable attribute. The name is a pointer/reference to an attribute within the OCSF event data. For example: file.name or src_endpoint.ip.`

Granted, this isn't a discrete column (e.g. `src_endpoint.ip` and `dst_endpoint.ip`) but the context is preserved as a value of the `name` column` allowing it to be in a query predicate.

It should be noted that these tables will be sparsely populated in most cases with empty cells as not all attributes are populated for each OCSF class. In a relational database, which is row-oriented, a sparsely populated table can waste space due to row-oriented fixed storage extents. This is not an issue with Parquet encoded columnar storage as the missing attributes (cells) don’t use any storage as they usually do with relational database storage extents. Tables may have many columns, but they are storage efficient.

However, minimizing the number of identical columns across tables may provide better compressibility, especially for row group columns of lower volume event types. This means when pre-joining multiple classes into a table, factoring out the identical attributes that become columns, since any single event will only populate the columns of a single OCSF class.

> ### Parquet

Reviewer 3: The parquet section will benefit with a few considerations about mapping OCSF data types should map to Parquet physical/logical types. This is a very real concern for any implementor, and guidance here should help.

Essentially, a OCSF type → Parquet type mapping table showing how boolean_t, integer_t, long_t, float_t, string_t, timestamp_t, json_t map to Parquet physical types would be great.

Let me know if you need help building that

Author: Sure, would be happy to include that if you can produce some examples.

> Parquet is a column-oriented file format whose files hold column values contiguously, unlike row-oriented file formats like CSV or JSON, where each file stores columns across rows, where rows are stored contiguously. Due to the diversity of column types, and distribution of column values, this traditional approach does not compress as efficiently, as values need to be skipped across the columns of each row. For low cardinality columns, the compressibility is highest. Just as important, search performance tends to favor contiguously stored column values since most projections and predicates do not require all the columns in a row. Therefore I/O load is greatly reduced, and the time required to return a resultset is optimized.

Reviewer 3: separately, nit, feel free to ignore. The line `"unlike row-oriented file formats like CSV or JSON, where each file stores columns across rows, where rows are stored contiguously."` can be simplified?

Author: Reworded this.

"Traditional row-oriented file formats like CSV or JSON, and RDBMS store columns across rows, and rows are stored contiguously along with all the column values."

### Observable and Category Pros
-	Many cybersecurity detections use cases can run against a single table across all product event sources without joins
-	Single product searches are possible against a single table for all products
-	Observables table has a fixed number of non-duplicative columns with many rows for optimum compression
-	Ideal for threat intelligence matching and retrospective matching against IOCs
-	Drill-down to details is efficient
-	Retains all the Pros of By Category 

### Observable and Category Cons
-	Requires an ETL that can surface the Observables, or product native sources that send Observables
-	Duplication of data: an additional table is required with duplicate values from the category dimension tables (since dimension tables hold full fidelity events)
-	ETL is less straightforward as the Observables array needs to be retrieved and parsed before events are stored (and the array removed from the events to reduce but not eliminate duplication in dimension tables).

## By Required Only

> This a simplified variant of the aforementioned Single Table Join-Union approach where only OCSF *Required* attributes are stored in one table. The number of columns and therefore Iceberg Field IDs is much smaller than in the Single-table model (< 1000 as of vs. 1.8), negating most of the cons of that approach.

Reviewer 3: typo in `< 1000 as of vs. 1.8` , assuming it should be `< 1000 as of v1.8`

Author: Fixed.

# OCSF – Iceberg Storage Strategies

> Paul Agbabian, April 2026; Updated May, June 2026

Reviewer 2: ```suggestion
Paul Agbabian, April 2026; Updated May, June, July 2026
```
(The July additions deserve a byline update.)

Author: Updated - added Rajas as well as contributor.

Contributions by Matthias Vallentin, Hunter Madison

## Overview
There are multiple ways that OCSF structured events can be stored for analysis in Parquet/Iceberg table format. The following list is not exhaustive but forms the most common approaches seen so far in the industry. 

1. Single table
2. By event source
3. By event category
4. By event class
5. By Observable and category
6. By Required only

> 7. Semi-structured payload

Reviewer 2: ```suggestion
5. By Observable and Category
6. By Required Only
7. Semi-structured Payload
```
(Capitalization to match the section titles below.)

Author: Done.

However, minimizing the number of identical columns across tables may provide better compressibility, especially for row group columns of lower volume event types. This means when pre-joining multiple classes into a table, factoring out the identical attributes that become columns, since any single event will only populate the columns of a single OCSF class.

In Iceberg/Parquet, a column absent from a data file costs zero (readers return null), null values compress to almost nothing via RLE-encoded definition levels, and adding columns is a pure metadata operation. The real cost of very wide tables is on the metadata path, which the Limits section below will cover.

### Parquet

> Parquet is a column-oriented file format whose files hold column values contiguously. Traditional row-oriented file formats like CSV or JSON, and RDBMS store columns across rows, and rows are stored contiguously along with all the column values. Due to the diversity of column types, and distribution of column values, this traditional approach does not compress as efficiently, as values need to be skipped across the columns of each row. For low cardinality columns, the compressibility is highest. Just as important, search performance tends to favor contiguously stored column values since most projections and predicates do not require all the columns in a row. Therefore I/O load is greatly reduced, and the time required to return a resultset is optimized.

Reviewer 2: Following up on @floydtree's ask for type mapping guidance, here is a proposed subsection, written against where the spec is going (Iceberg V3) rather than the lowest common denominator of today's engines:

### Type Mapping

OCSF defines a small set of primitive data types plus formatted subtypes that constrain strings and integers (`ip_t`, `mac_t`, `port_t`, ...). Each needs an Iceberg type, and two principles guide the choice: preserve fidelity with the normative JSON form, and prefer the type that unlocks engine features such as partition transforms, min/max pruning, and dictionary encoding.

| OCSF type | Iceberg type | Notes |
| --- | --- | --- |
| `boolean_t` | `boolean` | |
| `integer_t`, `port_t`, enum `*_id` | `int` | Low-cardinality enums dictionary-encode extremely well |
| `long_t` | `long` | |
| `float_t` | `double` | OCSF does not specify precision; `double` avoids silent loss |
| `string_t` and subtypes (`hostname_t`, `email_t`, `url_t`, `file_hash_t`, ...) | `string` | Normalize case at ingest so equality predicates and min/max stats behave |
| `timestamp_t` | `timestamptz` (microseconds) | See below |
| `datetime_t` | omit | Derived, human-readable sibling of `timestamp_t`; reconstruct at read time |
| `ip_t`, `subnet_t` | `string` | Iceberg has no IP type and, unlike Arrow, no extension type mechanism; canonicalize (RFC 5952 for IPv6) and accept that textual min/max bounds prune IP ranges poorly |
| `mac_t` | `string` | Canonicalize case and separators |
| `uuid_t` | `uuid` | 16-byte fixed on disk; core spec type |
| `bytestring_t` | `binary` | Decode from the base64 wire form |
| `json_t` | `variant` (V3) | The escape hatch for `unmapped` and `data`; JSON string is the legacy fallback |
| object | `struct` | Keep nested (see Structured Columns) |
| array | `list` | |

**The `timestamp_t` mapping is the consequential one.** OCSF encodes timestamps as epoch milliseconds, but that is a wire-format detail, not their semantics; the schema itself acknowledges this by offering `datetime_t` as a human-readable sibling. Storing them as `bigint` forfeits Iceberg's entire time machinery: the `year`/`month`/`day`/`hour` partition transforms are only defined for date/timestamp source columns, and engines treat longs as opaque numbers in time-range predicates. Converting to `timestamptz` at microsecond precision is lossless and enables the hidden partitioning this article recommends. Producers that need the normative form back can re-derive epoch milliseconds trivially.

**Enum siblings**: the optional `*_name` companions of `*_id` attributes are derived data. Omitting them and reconstructing labels at query time saves columns across every strategy in this article.

Reviewer 3: Thanks @mavam! That's pretty much what I had in mind. I initially considered including a Parquet physical encoding column as well, but since Iceberg deterministically maps to its underlying file formats (be it Parquet, Avro, ORC, etc), keeping the table at the Iceberg level makes sense.

@pagbabian-splunk Here's the table I'd recommend, keeping it specific to OCSF data types, rather than mixing in attributes as well. 

| OCSF Type | Iceberg Type | Notes |
|-----------|-------------|-------|
| `boolean_t` | `boolean` | |
| `integer_t` | `int` | Covers enums, counts, IDs. |
| `long_t` | `long` | Byte counts, durations, `type_uid`, etc. |
| `float_t` | `double` | CVSS scores, lat/lon, percentiles, etc. `double` matches JSON number wire representation |
| `timestamp_t` | `timestamptz` | Epoch millis × 1000. Enables `days()`/`hours()` partition transforms. Lower 3 micros digits always zero for OCSF-native events |
| `datetime_t` | `string` (optional) | RFC-3339 sibling of `timestamp_t`. May be omitted if storage efficiency is prioritized and timezone offset is not needed; reconstruct at query time via `from_unixtime()` |
| `port_t` | `int` | Range 0–65535 |
| `string_t` | `string` | All constrained subtypes below are physically identical on disk |
| `ip_t` | `string` | No native IP type in Iceberg. Canonicalize at ingest (RFC 5952 for IPv6). Min/max stats won't prune IP ranges; use bloom filters for exact-match |
| `subnet_t` | `string` | CIDR notation |
| `mac_t` | `string` | Canonicalize case and separators |
| `hostname_t` | `string` | |
| `email_t` | `string` | |
| `url_t` | `string` | |
| `file_path_t` | `string` | |
| `uuid_t` | `uuid` | 16-byte fixed on disk; core spec type |
| `bytestring_t` | `binary` | Decode base64 at ingest |
| `json_t` | `variant` (V3) / `string` (fallback) | The escape hatch for `unmapped` and `data`; JSON string is the legacy fallback |
| `object` | `struct` | Keep nested (see Structured Columns). Note that OCSF nests deeply; verify engine depth limits |
| `is_array: true` | `list` | |

Reviewer: I would leave `timestamp_t` as `long` and make `datetime_t` as `timestampz`. This means if a production is setting both `timestamp_t` as a long and  `datetime_t` ISO 8601 string, it's more likely that you can merge their OCSF events into your pipeline without any additional transformation or code changes since many engines will convert 8601 string to timestamp, but not long to timestamp.

Reviewer 2: @floydtree agreed on keeping the table at the Iceberg level; the physical Parquet encoding follows deterministically from it. Your expanded version works for me.

@hmadison I see the appeal, but I'd push back on making `datetime_t` the carrier of the timestamp type:

1. **`datetime_t` is optional; `timestamp_t` is normative.** The `_dt` siblings are opt-in, so a table anchored on `datetime_t` has no usable time column (and no partition source) for producers that emit only `time`. Partitioning and pruning have to hang off the required field.

2. **Schema-aware writing is already assumed.** An Iceberg writer can't be schema-oblivious anyway: this same table has it decoding base64 into `binary` and shredding objects into `struct`s. Epoch millis → `timestamptz` is a deterministic cast in the same pass. Arguably the cheapest transform in the whole mapping. Ingest "without any code changes" isn't really on the table once you're writing typed Parquet.

3. **The cost asymmetry favors ingest-time conversion.** Keeping `timestamp_t` as `long` saves one cast at write time but taxes every read: no `year()`/`month()`/`day()`/`hour()` partition transforms on the required column, opaque longs in time-range predicates, manual conversion in every query. Write-once, read-many argues for paying once at ingest.

If engine coercion of ISO 8601 strings is the draw, it points to the same place: coerce whichever wire form you have (`_dt` string or epoch millis) into a single `timestamptz` column at write time, and drop the redundant sibling.

Author: I will stay with `timestamptz` as this is needed for hidden partitioning which was discussed explicitly in a subsection.

### Structured Columns

OCSF makes heavy use of objects, which are structured sets of scalar and other object attributes. These attributes can directly become Iceberg structured columns (also encoded in Parquet files). While this can reduce the number of table columns, the underlying Iceberg files hold the elements of arrays and structures as scalar columns on disk. Although Parquet can store structures and lists without flattening them, in Iceberg each structure or array is decomposed (shredded) into its separate scalar column names assigned unique Field IDs. Elements of structures or arrays (or arrays of structures) can be accessed in queries via dot notation (or bracket and dot notation in Spark SQL).

#### Example Table

| Attribute | Iceberg Type | Comment |
| --------- | ------------ | ------- |

> | `time` | `bigint` | Base required, partition on `days(time)` |

Reviewer 2: With `time` as `bigint`, the `days(time)` partition spec in this row is not actually expressible: Iceberg's `day` transform is only defined for date/timestamp source columns. Mapping OCSF epoch milliseconds to microsecond `timestamptz` is lossless and makes the hidden partitioning work as described:
```suggestion
| `time` | `timestamptz` | Base required, partition on `days(time)` |
```

Reviewer 3: Given we are adding a dedicated table for OCSF Data types to Iceberg types, we should omit this from this table and keep it simple.

Author: If we want to have the example table in the document, we have to include the `time` column so I don't believe we should omit this. I do need to fix the Iceberg data type `bigint` from the original table.

Column IDs can run into internally reserved IDs for example. Query engines like AWS Athena can fail with more than 7500 columns, it has been reported. Statistics are collected for a limited number of columns, and configuring a high number of columns for statistics increases metadata size and impacts query planning. And while structured type columns can help keep top level columns in check, the depth and breadth of those structures can have practical limits, although they are quite high.

While these limits (e.g. 10,000 columns or 1000 structured sub-columns) might seem very high, and they are, one should realize that the combinatorial total of every OCSF attribute in all of its object, profile and class combinations can exceed one million scalar columns. Therefore, particular implementations of Iceberg with the various query engines in practice determine the actual limitations.

> Note per-file min/max stats (collected for the first 100 columns by default, write.metadata.metrics.max-inferred-column-defaults), impact manifest size, and planning time.

Reviewer 2: The comma before "impact" separates the subject from its verb, and the property name reads easier as code:
```suggestion
Note that per-file min/max stats (collected for the first 100 columns by default, controlled by `write.metadata.metrics.max-inferred-column-defaults`) impact manifest size and planning time.
```
This also overlaps with "Statistics are collected for a limited number of columns..." two paragraphs up; consider folding the two together.

Author: Done.

Further, classes in each category share common attributes, and all classes share common Base class attributes. (Note not all attributes are populated by every event source that uses the classes). Reuse of category-based tables across product event sources are by design. When new product event sources are onboarded, new tables are not generally required.

Because of the commonality across event categories, columns are not duplicated across the classes. Lastly, because the number of categories grows slowly across OCSF evolution, the number of tables will be relatively stable.

### Category Pros
-	A stable number of tables
-	Reuse of tables by multiple products
-	Minimize duplication of related columns within a table
-	Table holds all related classes in a domain
-	Reduce the number of joins within a domain’s category
-	ETL is straightforward as category_uid directs the events to its destination

### Category Cons
-	Searching for all events from a single product requires searching all tables if the product events span categories

> -   Table will be larger than by source or by event class impacting compaction

Reviewer 2: Carrying @hmadison's point from the Single Table cons over here for consistency: a larger table constrains partitioning rather than compaction, since unioning Parquet files is cheap.
```suggestion
-   Table will be larger than by source or by event class impacting partitioning
```

Author: Done.

This strategy is hybrid. It takes advantage of an OCSF schema feature called Observables.

Observables are an array of the most common attributes surfaced from any of the events that are emitted in a flat generic array. Each element of the array is a single Observable whose `value` attribute holds the element data. The `name` attribute provides the schema context of the attribute, for example `src_endpoint.ip` and `dst_endpoint.ip` are distinct attributes of the `ip` observable, and the `class_uid` and `activity_id` are present for behavior.

The original intent of the observables was for threat intelligence matching, where the total fidelity and context of the event was not required, just the observable values that might match Indicators of Compromise (IoCs).

There is another use case for Observables: a table constructed for observables will be common across every event class, and therefore every event across all products. This table can have foreign keys to dimension tables much like a STAR schema in OLAP. The Observables array is flattened into about 40 columns, along with the most important Base event attribute columns that identify and classify the events.

> In practice, many analytics can run very efficiently directly against a single table across all products and classes, while drill-down and detailed investigation is performed by a minimum number of joins to dimension tables. Value-match detections, such as IOC matching and retrospective hunting over observable values, need no joins at all; the specific attribute is available in the `name` attribute (e.g. `src_endpoint.ip` or `file.name`) and is available to a query predicate. Detections that depend on richer event context than the observables and Base attributes carry will still join to the dimension tables.

Reviewer 2: Small wording pass: "is available ... and is available" doubles up, and it's worth being precise that `name` carries the schema context while `value` carries the data:
```suggestion
In practice, many analytics can run very efficiently directly against a single table across all products and classes, while drill-down and detailed investigation is performed by a minimum number of joins to dimension tables. Value-match detections, such as IOC matching and retrospective hunting over observable values, need no joins at all; the `name` attribute identifies the specific attribute (e.g. `src_endpoint.ip` or `file.name`) and the `value` attribute is directly available to a query predicate. Detections that depend on richer event context than the observables and Base attributes carry will still join to the dimension tables.
```

Author: Done.

### Observable and Category Pros
-	Many cybersecurity detections use cases can run against a single table across all product event sources without joins
-	Single product searches are possible against a single table for all products
-	Observables table has a fixed number of non-duplicative columns with many rows for optimum compression
-	Ideal for threat intelligence matching and retrospective matching against IOCs
-	Drill-down to details is efficient
-	Retains all the Pros of By Category 

### Observable and Category Cons
-	Requires an ETL that can surface the Observables, or product native sources that send Observables
-	Duplication of data: an additional table is required with duplicate values from the category dimension tables (since dimension tables hold full fidelity events)
-	ETL is less straightforward as the Observables array needs to be retrieved and parsed before events are stored (and the array removed from the events to reduce but not eliminate duplication in dimension tables).

## By Required Only

> This a simplified variant of the aforementioned Single Table Join-Union approach where only OCSF *Required* attributes are stored in one table. The number of columns and therefore Iceberg Field IDs is much smaller than in the Single-table model (< 1000 as of v1.8), negating most of the cons of that approach.

Reviewer 2: ```suggestion
This is a simplified variant of the aforementioned Single Table Join-Union approach where only OCSF *Required* attributes are stored in one table. The number of columns and therefore Iceberg Field IDs is much smaller than in the Single-table model (< 1000 as of v1.8), negating most of the cons of that approach.
```
(Missing "is".)

Author: Done.

### Required Only Pros
-	A stable number of tables (1)
-	Reuse of table by multiple products

> -	Minimize duplication of related columns within a table

Reviewer 2: This pro carries over from Single Table, but with only required attributes stored there is little column duplication left to minimize; suggest dropping it.
```suggestion
```

Author: Done.

-   No cross-table joins for all use cases including single source or product queries
-   Single table for all ETL targets
-   Leverage schema evolution for new class and partition maintenance

### Required Only Cons

> -   Full event fidelity is not available

Reviewer 2: Worth spelling out the consequence: without full fidelity, drill-down has to go somewhere else, so in practice this strategy is paired with raw event retention or one of the other strategies.
```suggestion
-   Full event fidelity is not available; drill-down requires a separate full-fidelity store
```

Author: Done.

-   All use cases depend on a single table with cross-use-case data mingled in files and partitions
-   May not satisfy a general case of applications due to an arbitrary (but necessary) column selection

## Semi-structured Payload

> This strategy can be combined with any of the other strategies. It promotes a small, stable set of attributes to real columns and stores the complete event in a single semi-structured payload column. The promoted set consists of the Base event required attributes (time, class_uid, category_uid, activity_id, severity_id, metadata) plus a curated set of frequently queried attributes that will be use-case specific. The promoted columns carry partitioning, sorting, and column statistics; the payload column preserves full fidelity.

Reviewer 2: ```suggestion
This strategy can be combined with any of the other strategies. It promotes a small, stable set of attributes to real columns and stores the complete event in a single semi-structured payload column. The promoted set consists of the Base event required attributes (`time`, `class_uid`, `category_uid`, `activity_id`, `severity_id`, `metadata`) plus a curated set of frequently queried attributes that will be use-case specific. The promoted columns carry partitioning, sorting, and column statistics; the payload column preserves full fidelity.
```
(Backticks for the attribute names, matching the rest of the document.)

Author: Done.

| `datetime_t` | `string` (optional) | RFC-3339 sibling of `timestamp_t`. May be omitted if storage efficiency is prioritized and timezone offset is not needed; reconstruct at query time via `from_unixtime()` |
| `port_t` | `int` | Range 0–65535 |
| `string_t` | `string` | All constrained subtypes below are physically identical on disk |
| `ip_t` | `string` | No native IP type in Iceberg. Canonicalize at ingest (RFC 5952 for IPv6). Min/max stats won't prune IP ranges; use bloom filters for exact-match |
| `subnet_t` | `string` | CIDR notation |
| `mac_t` | `string` | Canonicalize case and separators |
| `hostname_t` | `string` | |
| `email_t` | `string` | |
| `url_t` | `string` | |
| `file_path_t` | `string` | |
| `uuid_t` | `uuid` | 16-byte fixed on disk; core spec type |
| `bytestring_t` | `binary` | Decode base64 at ingest |
| `json_t` | `variant` (V3) / `string` (fallback) | The escape hatch for `unmapped` and `data`; JSON string is the legacy fallback |
| `object` | `struct` | Keep nested (see Structured Columns). Note that OCSF nests deeply; verify engine depth limits |
| `is_array: true` | `list` | |

#### A Note on Time Columns and the OCSF Timestamp Logical Type

> The OCSF `Timetamp` (`timestamp_t`) data type is defined as the number of milliseconds stored as a `long` since the Unix Epoch, 01/01/1970 00:00:00 UTC. The Iceberg `timestamptz` type is defined as the number of *microseconds* since the Unix Epoch or in the case of `timestamptz_ns`, *nanoseconds*. (The `time`, `timestamp` and `timestamp_ns` types are not UTC but local time). Therefore, when converting from OCSF `Timestamp` values to Iceberg `timestamptz` columns, you should multiply by a factor of 1,000 (or 1,000,000 for nanoseconds) in order to avoid time shift errors. Iceberg aware programming interfaces like Java or Spark will handle conversion based on their own time and date types, but for directly inserting integral data values this must be done explicitly.

Reviewer 2: ```suggestion
The OCSF `Timestamp` (`timestamp_t`) data type is defined as the number of milliseconds stored as a `long` since the Unix Epoch, 01/01/1970 00:00:00 UTC. The Iceberg `timestamptz` type is defined as the number of *microseconds* since the Unix Epoch or in the case of `timestamptz_ns`, *nanoseconds*. (The `timestamp` and `timestamp_ns` types are not UTC but local time; `time` is a time-of-day type.) Therefore, when converting from OCSF `Timestamp` values to Iceberg `timestamptz` columns, you should multiply by a factor of 1,000 (or 1,000,000 for nanoseconds) in order to avoid time shift errors. Iceberg-aware programming interfaces like Java or Spark will handle conversion based on their own time and date types, but for directly inserting integral data values this must be done explicitly.
```
("Timetamp" → "Timestamp"; `time` is a time-of-day type rather than a local-time timestamp; hyphenated "Iceberg-aware".)

Author: Thanks @mavam fixed this one.

Column IDs can run into internally reserved IDs for example. Query engines like AWS Athena can fail with more than 7500 columns, it has been reported. Statistics are collected for a limited number of columns, and configuring a high number of columns for statistics increases metadata size and impacts query planning. And while structured type columns can help keep top level columns in check, the depth and breadth of those structures can have practical limits, although they are quite high.

While these limits (e.g. 10,000 columns or 1000 structured sub-columns) might seem very high, and they are, one should realize that the combinatorial total of every OCSF attribute in all of its object, profile and class combinations can exceed one million scalar columns. Therefore, particular implementations of Iceberg with the various query engines in practice determine the actual limitations.

> Note that per-file min/max stats (collected for the first 100 columns by default, controlled by `write.metadata.metrics.max-inferred-column-defaults`), impact manifest size, and planning time.

Reviewer 2: ```suggestion
Note that per-file min/max stats (collected for the first 100 columns by default, controlled by `write.metadata.metrics.max-inferred-column-defaults`) impact manifest size and planning time.
```
(Drops the comma between subject and verb, carrying over the July 10 suggestion in full.)

Author: Fixed this one.
